fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java

```diff
@@ -0,0 +1,195 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.analysis.Expr;
+import com.starrocks.catalog.Function;
+import com.starrocks.catalog.FunctionSet;
+import com.starrocks.catalog.Type;
+import com.starrocks.common.Pair;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.operator.AggType;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+import org.jetbrains.annotations.NotNull;
+import org.spark_project.guava.collect.ImmutableMap;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/*
+ *
+ * Rewrite count(distinct xx) group by x when distinct columns satisfy scan distribution
+ *
+ * e.g.
+ * select count(distinct xx) from t group by x
+ * ->
+ * select count(xx) from (select x, xx from t group by x, xx) tt group by x;
+ *
+ */
+public class GroupByCountDistinctRewriteRule extends TransformationRule {
+    // multi-stage distinct function mapping
+    // origin aggregate -> 2nd stage
+    private static final Map<String, String> DISTINCT_FUNCTION_TRANS = ImmutableMap.<String, String>builder()
+            .put(FunctionSet.COUNT, FunctionSet.COUNT)
+            .put(FunctionSet.SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_COUNT, FunctionSet.COUNT)
+            .build();
+
+    // multi-stage other function mapping
+    // origin aggregate -> Pair<1st stage, 2nd stage>
+    // these aggregate function can be support multi-stage
+    private static final Map<String, Pair<String, String>> OTHER_FUNCTION_TRANS =
```

Comment:
Why need this map? keep DISTINCT_FUNCTION_TRANS isn't enough?

Reply:
DISTINCT_FUNCTION_TRANS don't need first stage